gateway/modules/crud/batcher.go

```diff
@@ -0,0 +1,125 @@
+package crud
+
+import (
+	"context"
+	"time"
+
+	"github.com/spaceuptech/space-cloud/gateway/model"
+	"github.com/spaceuptech/space-cloud/gateway/utils"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spaceuptech/space-cloud/gateway/config"
+)
+
+// data structure for storing a group of channels
+// against particular table/collection of particular database
+type (
+	batchMap map[string]map[string]map[string]batchChannels // keys are project name, dbAlias, table name
+
+	batchChannels struct {
+		request batchRequestChan
+		close   chan struct{}
+	}
+
+	batchRequestChan chan batchRequest
+
+	batchRequest struct {
+		document interface{}
+		response chan error
+	}
+)
+
+// CloseBatchOperation closes all go routines associated with individual collection for batch operation
+func (m *Module) CloseBatchOperation() {
+	for _, projectValue := range m.batchMapTableToChan {
+		for _, dbAliasValue := range projectValue {
+			for _, tableValue := range dbAliasValue {
+				tableValue.close <- struct{}{}
+			}
+		}
+	}
+}
+
+// initBatchOperation creates go routines for executing batch operation associated with individual collection
+func (m *Module) initBatchOperation(project string, crud config.Crud) {
+	batch := batchMap{}
+	for dbAlias, dbInfo := range crud {
+		if dbInfo.Enabled {
+			for tableName, tableInfo := range dbInfo.Collections {
+				done := make(chan struct{})                                          // channel for closing go routine
+				addInsertToBatchCh := make(batchRequestChan, tableInfo.BatchRecords) // channel for adding request to batch op
+				go m.insertBatchExecutor(done, addInsertToBatchCh, dbInfo.BatchTime, project, dbAlias, tableName, tableInfo.BatchRecords)
+				if batch[project] == nil {
+					batch[project] = map[string]map[string]batchChannels{dbAlias: {tableName: {request: addInsertToBatchCh, close: done}}}
+					continue
+				}
+				batch[project][dbAlias][tableName] = batchChannels{request: addInsertToBatchCh, close: done}
```

Comment:
What if `batch[project][dbAlias]` is `nil`? You should check for that as well.

Reply:
this condition will not occur as if a project doen't exists
the whole structure gets initialized for that project
next time for the same project the whole structure is already initialized
see the if statement above